src/main/java/com/artipie/asto/fs/FileStorage.java

```diff
@@ -154,23 +156,43 @@ public FileStorage(final Path path, final ExecutorService exec) {
     public CompletableFuture<Void> save(final Key key, final Content content) {
         return CompletableFuture.supplyAsync(
             () -> {
-                final Path file = this.path(key);
-                try {
-                    Files.createDirectories(file.getParent());
-                } catch (final IOException iex) {
-                    throw new UncheckedIOException(iex);
-                }
-                return file;
-            },
-            this.exec
+                final Path tmp = Paths.get(
+                    this.dir.toString(),
+                    String.format("%s.%s.tmp", key.string(), UUID.randomUUID())
+                );
+                tmp.getParent().toFile().mkdirs();
+                return tmp;
+            }
         ).thenCompose(
-            path -> new File(path).write(
+            tmp -> new File(tmp).write(
                 new OneTimePublisher<>(content),
                 this.exec,
                 StandardOpenOption.WRITE,
                 StandardOpenOption.CREATE,
                 StandardOpenOption.TRUNCATE_EXISTING
-            )
+            ).thenApply(
+                nothing -> {
+                    try {
+                        final Path target = this.path(key);
+                        target.getParent().toFile().mkdirs();
+                        Files.move(tmp, target, StandardCopyOption.REPLACE_EXISTING);
+                    } catch (final IOException ex) {
+                        throw new UncheckedIOException(ex);
+                    }
+                    return null;
+                }
+            ).handle(
+                (nothing, throwable) -> {
```

Comment:
@olegmoz add this lambda to a private method to make the code more readable

Reply:
@Vatavuk this lambda is closely coupled with `tmp` file creation, so I'd rather keep it here